### PR TITLE
Integration test: Verify warm COINBASE address (EIP-3651)

### PR DESCRIPTION
**Pull request by LuisPH3**
Verify warm COINBASE address [EIP-3651](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-3651.md)

The tests check that the address returned is Zero, and that the access to such address is considered warm. This is done by comparing the costs of cold/warm accesses.

In addition to COINBASE, the ORIGIN instruction is tested. 

The test is written with the following structure to minimize the number of test-nets initialized:
```
--- PASS: TestAddressAccess (12.23s)
    --- PASS: TestAddressAccess/coinbase_yields_zero_address (0.00s)
    --- PASS: TestAddressAccess/access_is_warm (2.13s)
        --- PASS: TestAddressAccess/access_is_warm/coinbase (0.64s)
        --- PASS: TestAddressAccess/access_is_warm/origin (0.74s)
        --- PASS: TestAddressAccess/access_is_warm/access_list (0.75s)
PASS
```
This way, the overhead of setting up an integration environment is paid only once for the complete file.

### CODE REVIEW

File: tests/coinbase_test.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	receipt, err = net.Apply(func(opts *bind.TransactOpts) (*types.Transaction, error) {
+		return contract.TouchAddress(opts, coinbaseAddress)
+	})
+	checkTxExecution(t, receipt, err)
+	coinBaseCost := receipt.GasUsed
+
+	// CoinBase address access has the same cost as warm access
+	if warmCost == coinBaseCost {
+		t.Errorf("Expected coinbase access to be cheaper than warm access")
+	}
```
Comment (HerbertJordan) [inline tests/coinbase_test.go:84]: Shouldn't this test `warmCost != coinBaseCost`?

**Comment by HerbertJordan** [comment]
> It has been noticed that COINBASE is not automatically created, and first invocation may add creation fees. This may be malfunction. wdyt?

What exactly do you mean with "automatically created" in this context?

File: tests/integration_test_net.go [hunk 2]
```diff
@@ -74,7 +74,7 @@ func StartIntegrationTestNet(directory string) (*IntegrationTestNet, error) {
 			"--http", "--http.addr", "0.0.0.0", "--http.port", "18545",
 			"--http.api", "admin,eth,web3,net,txpool,ftm,trace,debug",
 			"--ws", "--ws.addr", "0.0.0.0", "--ws.port", "18546", "--ws.api", "admin,eth,ftm",
-			"--pprof", "--pprof.addr", "0.0.0.0",
+			// "--pprof", "--pprof.addr", "0.0.0.0",
```
[the remark this answers is not in the document]
Comment (LuisPH3) [reply tests/integration_test_net.go:77]: sorry. i did not wanted to commit this line.

**Comment by LuisPH3** [comment]
> > It has been noticed that COINBASE is not automatically created, and first invocation may add creation fees. This may be malfunction. wdyt?
> 
> What exactly do you mean with "automatically created" in this context

I attributed some extra costs to the creation of the account at COINBASE address. The current tests do not show this behaviour.

**Review by facuMH — DISMISSED** [review]
Nice to see number that make sense after all the testing

File: tests/address_access_test.go
Comment (HerbertJordan) [inline tests/address_access_test.go:1]: This turned out very nice, thanks! I also like a lot the idea with the network reuse.

As a follow-up: the tests should be extend to also make sure that the coinbase is considered warm when running a transaction on the archive. You can send a regular transaction to a historic state by using the `client``s [CallContract](https://github.com/ethereum/go-ethereum/blob/8c73523812d3a4945bf9d9b7fea24785cfb441b2/ethclient/ethclient.go#L525) method.

**Review by HerbertJordan — APPROVED** [review]

**Review by jenikd — DISMISSED** [review]
Nice progress here. Thanks for that. Only one small note. Link to EIP in description is leading to a wrong EIP

**Comment by LuisPH3** [comment]
> Nice progress here. Thanks for that. Only one small note. Link to EIP in description is leading to a wrong EIP

fixed, thanks

File: tests/address_access_test.go
Comment (LuisPH3) [reply tests/address_access_test.go:1]: Right, I added some tests. View accesses on an existing net are really fast. 
```
--- PASS: TestAddressAccess (12.52s)
    --- PASS: TestAddressAccess/coinbase_yields_zero_address (0.00s)
    --- PASS: TestAddressAccess/tx_access_is_warm (2.17s)
        --- PASS: TestAddressAccess/tx_access_is_warm/coinbase (0.75s)
        --- PASS: TestAddressAccess/tx_access_is_warm/origin (0.65s)
        --- PASS: TestAddressAccess/tx_access_is_warm/access_list (0.76s)
    --- PASS: TestAddressAccess/archive_access_is_warm (0.00s)
        --- PASS: TestAddressAccess/archive_access_is_warm/origin (0.00s)
        --- PASS: TestAddressAccess/archive_access_is_warm/coinbase (0.00s)
```

**Review by facuMH — APPROVED** [review]

**Review by HerbertJordan — APPROVED** [review]

**Pull request merged**